File: tools/samples/sample.go
package members

import (
	"context"
	"errors"
	"fmt"
)

type Role string

const (
	RoleOwner  Role = "owner"
	RoleEditor Role = "editor"
	RoleViewer Role = "viewer"
)

var ErrNotFound = errors.New("member not found")

type Member struct {
	ID    string `json:"id"`
	Email string `json:"email"`
	Role  Role   `json:"role"`
}

type Repository interface {
	Find(ctx context.Context, id string) (*Member, error)
	Save(ctx context.Context, m *Member) error
}

// Promote raises a member's role, leaving owners untouched.
func Promote(ctx context.Context, repo Repository, id string, to Role) (*Member, error) {
	m, err := repo.Find(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("find %s: %w", id, err)
	}
	if m.Role == RoleOwner && to != RoleOwner {
		return m, nil
	}
	m.Role = to
	return m, repo.Save(ctx, m)
}

File: tools/samples/sample.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    Owner,
    Editor,
    Viewer,
}

#[derive(Debug, Clone)]
pub struct Member {
    pub id: String,
    pub email: String,
    pub role: Role,
}

#[derive(Debug, thiserror::Error)]
pub enum PromoteError {
    #[error("no member {0}")]
    NotFound(String),
}

impl Member {
    /// Raises the member's role, leaving owners untouched.
    pub fn promote(&mut self, to: Role) -> &Self {
        match (self.role, to) {
            (Role::Owner, r) if r != Role::Owner => self,
            _ => {
                self.role = to;
                self
            }
        }
    }
}

pub fn promote(cache: &mut HashMap<String, Member>, id: &str) -> Result<Member, PromoteError> {
    let member = cache
        .get_mut(id)
        .ok_or_else(|| PromoteError::NotFound(id.into()))?;
    Ok(member.promote(Role::Editor).clone())
}
